jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -212,51 +207,44 @@ public boolean tryExecute(Runnable task)
     {
         if (LOG.isDebugEnabled())
             LOG.debug("{} tryExecute {}", this, task);
-
         if (task == null)
             return false;
 
-        ReservedThread thread = _stack.pollFirst();
-        if (thread == null)
-        {
-            if (task != STOP)
-                startReservedThread();
-            return false;
-        }
-
-        int size = _size.decrementAndGet();
-        if (!thread.offer(task))
-            return false;
-
+        boolean offered = _queue.offer(task);
+        int size = offered ? _count.addAndGetLo(-1) : _count.getLo();
         if (size == 0 && task != STOP)
             startReservedThread();
-
-        return true;
+        return offered;
     }
 
     private void startReservedThread()
     {
-        try
+        while (true)
         {
-            while (true)
+            long count = _count.get();
+            int pending = AtomicBiInteger.getHi(count);
+            int size = AtomicBiInteger.getLo(count);
+            if (size == Integer.MAX_VALUE || pending + size >= _capacity)
+                return;
+
+            if (size == 0)
+                _lastZero = System.nanoTime();
+            if (!_count.compareAndSet(count, pending + 1, size))
+                continue;
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} startReservedThread p={}", this, pending + 1);
+            try
             {
-                // Not atomic, but there is a re-check in ReservedThread.run().
-                int pending = _pending.get();
-                int size = _size.get();
-                if (pending + size >= _capacity)
-                    return;
-                if (_pending.compareAndSet(pending, pending + 1))
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} startReservedThread p={}", this, pending + 1);
-                    _executor.execute(new ReservedThread());
-                    return;
-                }
+                ReservedThread thread = new ReservedThread();
+                addBean(thread);
```

Comment:
`addBean` ends up modifying a COW collection, that's going to generate a lot of garbage on a busy server where lots of reserved threads are created.

Reply:
It is a worry, but this is not the fine grained reserved thread add/create, only the courser grained lifecycle.   A thread can become reserved, run a task, become reserved, run a task, etc. many times without touching the COW collection.    Note that means the collection contains all the reserved threads, plus all the threads that were once reserved but are not RUNNING.

Reply:
When there are many more threads calling tryExecute than there are reserved threads in the pool, most reserved threads are thrown away after they're done with their execution because the pending counter is constantly pegged at the maximum.

I'm not sure how realistic such load is, but it's very easy to write a synthetic test that triggers this behavior.

Reply:
I'm not sure that is true.  If there are many more calls to tryExecute than the capacity of the reserved thread pool, then most calls return false and nothing is done.   A reserved thread is only started if there is capacity for it.  It is only stopped if it goes idle.   So modifications should not be too frequent.... but I'll have a look at the idle mechanism and make sure we don't kill all our reserved threads at once.

Reply:
I'm using a concurrent queue now for holding the threads for dumping